.tree-sitter-lint/tree-sitter-lint-local/src/lib.rs

```diff
@@ -1 +1,126 @@
+    FromFileRunContextProvidedTypesOnceLockStorage, MutRopeOrSlice, Plugin, Rule,
+    ViolationWithContext,
+};
+
+pub fn run_and_output() {
+    tree_sitter_lint::run_and_output(
+        args_to_config(Args::parse()),
+        FromFileRunContextInstanceProviderFactoryLocal,
+    );
+}
+
+pub fn run_for_slice<'a>(
+    file_contents: impl Into<RopeOrSlice<'a>>,
+    tree: Option<&Tree>,
+    path: impl AsRef<Path>,
+    args: Args,
+    language: SupportedLanguage,
+) -> Vec<ViolationWithContext> {
+    tree_sitter_lint::run_for_slice(
+        file_contents,
+        tree,
+        path,
+        args_to_config(args),
+        language,
+        &FromFileRunContextInstanceProviderFactoryLocal,
+    )
+}
+
+pub fn run_fixing_for_slice<'a>(
+    file_contents: impl Into<MutRopeOrSlice<'a>>,
+    tree: Option<&Tree>,
+    path: impl AsRef<Path>,
+    args: Args,
+    language: SupportedLanguage,
+) -> Vec<ViolationWithContext> {
+    tree_sitter_lint::run_fixing_for_slice(
+        file_contents,
+        tree,
+        path,
+        args_to_config(args),
+        language,
+        &FromFileRunContextInstanceProviderFactoryLocal,
+    )
+}
+
+struct LocalLinterConcrete;
+
+impl LocalLinter for LocalLinterConcrete {
+    fn run_for_slice<'a>(
+        &self,
+        file_contents: impl Into<RopeOrSlice<'a>>,
+        tree: Option<&Tree>,
+        path: impl AsRef<Path>,
+        args: Args,
+        language: SupportedLanguage,
+    ) -> Vec<ViolationWithContext> {
+        run_for_slice(file_contents, tree, path, args, language)
+    }
+
+    fn run_fixing_for_slice<'a>(
+        &self,
+        file_contents: impl Into<MutRopeOrSlice<'a>>,
+        tree: Option<&Tree>,
+        path: impl AsRef<Path>,
+        args: Args,
+        language: SupportedLanguage,
+    ) -> Vec<ViolationWithContext> {
+        run_fixing_for_slice(file_contents, tree, path, args, language)
+    }
+}
+
+pub async fn run_lsp() {
+    lsp::run(LocalLinterConcrete).await;
+}
+
+fn args_to_config<T: FromFileRunContextInstanceProviderFactory>(args: Args) -> Config<T> {
+    args.load_config_file_and_into_config(all_plugins(), all_standalone_rules())
+}
+
+fn all_plugins<T: FromFileRunContextInstanceProviderFactory>() -> Vec<Plugin<T>> {
+    vec![tree_sitter_lint_plugin_replace_foo_with::instantiate()]
+}
+
+fn all_standalone_rules<T: FromFileRunContextInstanceProviderFactory>() -> Vec<Arc<dyn Rule<T>>> {
+    local_rules::get_rules()
+}
+
+struct FromFileRunContextInstanceProviderFactoryLocal;
+
+impl FromFileRunContextInstanceProviderFactory for FromFileRunContextInstanceProviderFactoryLocal {
+    type Provider<'a> = FromFileRunContextInstanceProviderLocal<'a>;
+
+    fn create<'a>(&self) -> Self::Provider<'a> {
+        FromFileRunContextInstanceProviderLocal {
+            tree_sitter_lint_plugin_replace_foo_with_provided_instances:
+                tree_sitter_lint_plugin_replace_foo_with::ProvidedTypes::<'a>::once_lock_storage(),
```

Comment:
Ok so the idea is that every single plugin will be expected to "expose" a `pub type ProvidedType<'a> = (SomeProvidedType<'a>, SomeOtherProvidedType<'a>);`

It's unfortunate that that can't get "type-checked" on the plugin side but I didn't think of a way to achieve that - I tried eg changing `Plugin` to a trait and making this an associated type on that plugin's `Plugin` implementation, but then those `Arc<dyn Plugin>`'s can't be stored in a `Vec` because they might have different associated types (so are of different types)

So basically I couldn't see a way to get a "unified" `Plugin` type/trait to be able to "expose" a set of provided types

So went with it being a separate "declaration"

So what this is doing (thinking in terms of code-generating this) is looping through all of the plugins and creating a struct field (in our custom local "instance provider"-trait-implementing struct) for each one whose type is (per above comment) a "pre-allocated" `OnceLock`-based "storage" for all of that plugin's declared provided types

So that sort of "adds up to" a "pre-allocated" slot for every "valid" (ie declared by some plugin that we're using) provided type